src/modules/rover_mecanum/RoverMecanumControl/RoverMecanumControl.cpp

```diff
@@ -0,0 +1,220 @@
+			   _param_rm_p_gain_yaw.get(),  // Proportional gain
+			   _param_rm_i_gain_yaw.get(),  // Integral gain
+			   0.f,  // Derivative gain
+			   _max_yaw_rate,  // Integral limit
+			   _max_yaw_rate);  // Output limit
+}
+
+void RoverMecanumControl::computeMotorCommands(const float vehicle_yaw, const float vehicle_yaw_rate,
+		const float vehicle_forward_speed, const float vehicle_lateral_speed)
+{
+	// Timestamps
+	hrt_abstime timestamp_prev = _timestamp;
+	_timestamp = hrt_absolute_time();
+	const float dt = math::constrain(_timestamp - timestamp_prev, 1_ms, 5000_ms) * 1e-6f;
+
+	// Update mecanum setpoint
+	_rover_mecanum_setpoint_sub.update(&_rover_mecanum_setpoint);
+
+	// Closed loop yaw control
+	if (PX4_ISFINITE(_rover_mecanum_setpoint.yaw_setpoint)) {
+		const float heading_error = matrix::wrap_pi(_rover_mecanum_setpoint.yaw_setpoint - vehicle_yaw);
+		_rover_mecanum_setpoint.yaw_rate_setpoint = pid_calculate(&_pid_yaw, heading_error, 0, 0, dt);
+
+	} else {
+		pid_reset_integral(&_pid_yaw);
+	}
+
+	// Yaw rate control
+	float speed_diff_normalized{0.f};
+
+	if (PX4_ISFINITE(_rover_mecanum_setpoint.yaw_rate_setpoint)) { 	// Closed loop yaw rate control
+		if (_param_rm_wheel_track.get() > FLT_EPSILON && _param_rm_max_thr_yaw_r.get() > FLT_EPSILON) { // Feedforward
+			const float speed_diff = _rover_mecanum_setpoint.yaw_rate_setpoint * _param_rm_wheel_track.get();
+			speed_diff_normalized = math::interpolate<float>(speed_diff, -_param_rm_max_thr_yaw_r.get(),
+						_param_rm_max_thr_yaw_r.get(), -1.f, 1.f);
+		}
+
+		speed_diff_normalized = math::constrain(speed_diff_normalized +
+							pid_calculate(&_pid_yaw_rate, _rover_mecanum_setpoint.yaw_rate_setpoint, vehicle_yaw_rate, 0, dt),
+							-1.f, 1.f); // Feedback
+
+	} else { // Use normalized setpoint
+		speed_diff_normalized = PX4_ISFINITE(_rover_mecanum_setpoint.yaw_rate_setpoint_normalized) ? math::constrain(
+						_rover_mecanum_setpoint.yaw_rate_setpoint_normalized, -1.f, 1.f) : 0.f;
+	}
+
+	// Speed control
+	float forward_throttle{0.f};
+	float lateral_throttle{0.f};
+
+	if (PX4_ISFINITE(_rover_mecanum_setpoint.forward_speed_setpoint)
+	    && PX4_ISFINITE(_rover_mecanum_setpoint.lateral_speed_setpoint)) { // Closed loop speed control
+		// Closed loop forward speed control
+		if (_param_rm_max_thr_spd.get() > FLT_EPSILON) { // Feedforward
+			forward_throttle = math::interpolate<float>(_rover_mecanum_setpoint.forward_speed_setpoint,
+					   -_param_rm_max_thr_spd.get(), _param_rm_max_thr_spd.get(), -1.f, 1.f);
+		}
+
+		forward_throttle += pid_calculate(&_pid_forward_throttle, _rover_mecanum_setpoint.forward_speed_setpoint,
+						  vehicle_forward_speed, 0, dt); // Feedback
+
+		// Closed loop lateral speed control
+		if (_param_rm_max_thr_spd.get() > FLT_EPSILON) { // Feedforward
+			lateral_throttle += math::interpolate<float>(_rover_mecanum_setpoint.lateral_speed_setpoint,
+					    -_param_rm_max_thr_spd.get(), _param_rm_max_thr_spd.get(), -1.f, 1.f);
```

Comment:
This assumes a symmetric rover that has the same throttle <--> speed mapping for body x and y axis right?

Reply:
Yes, that is the case for rovers using four $45\degree$ mecanum wheels, which seems to be the standard.
(Different configuartions would also have different inverse kinematics).

Reply:
ah I see, let's keep it like that for now